libs/lb/lb-rs/src/rpc.rs

```diff
@@ -0,0 +1,200 @@
+use crate::dispatch::dispatch;
+use crate::model::errors::core_err_unexpected;
+use crate::{LbResult, LbServer};
+use bincode;
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddrV4;
+use std::sync::Arc;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpListener, TcpStream};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u16)]
+pub enum Method {
+    CreateAccount = 0,
+    ImportAccount,
+    ImportAccountPrivateKeyV1,
+    ExportAccountPrivateKey,
+    ExportAccountPrivateKeyV1,
+    ExportAccountPrivateKeyV2,
+    ExportAccountPhrase,
+    ExportAccountQr,
+    DeleteAccount,
+    SuggestedDocs,
+    DisappearAccount,
+    DisappearFile,
+    ListUsers,
+    GetAccountInfo,
+    ValidateAccount,
+    ValidateServer,
+    FileInfo,
+    RebuildIndex,
+    BuildIndex,
+    SetUserTier,
+    UpgradeAccountStripe,
+    UpgradeAccountGooglePlay,
+    UpgradeAccountAppStore,
+    CancelSubscription,
+    GetSubscriptionInfo,
+    DebugInfo,
+    ReadDocument,
+    WriteDocument,
+    ReadDocumentWithHmac,
+    SafeWrite,
+    Subscribe,
+    CreateFile,
+    RenameFile,
+    MoveFile,
+    Delete,
+    Root,
+    ListMetadatas,
+    GetChildren,
+    GetAndGetChildrenRecursively,
+    GetFileById,
+    LocalChanges,
+    ImportFiles,
+    ExportFile,
+    ExportFileRecursively,
+    TestRepoIntegrity,
+    GetAccount,
+    CreateLinkAtPath,
+    CreateAtPath,
+    GetByPath,
+    GetPathById,
+    ListPaths,
+    ListPathsWithIds,
+    ShareFile,
+    GetPendingShares,
+    RejectShare,
+    CalculateWork,
+    Sync,
+    GetLastSyncedHuman,
+    GetTimestampHumanString,
+    GetUsage,
+    GetUncompressedUsageBreakdown,
+    GetUncompressedUsage,
+    Search,
+    Status,
+    GetConfig,
+    GetLastSynced,
+    GetSearch,
+    GetKeychain,
+}
+
+pub async fn call_rpc<T, R>(socket_address: SocketAddrV4, method: Method, args: T) -> LbResult<R>
+where
+    T: Serialize,
+    R: for<'de> Deserialize<'de>,
+{
+    let mut stream = TcpStream::connect(socket_address)
+        .await
+        .map_err(core_err_unexpected)?;
+
+    send_rpc_request(&mut stream, method, &args).await?;
+    recv_rpc_response(&mut stream).await
+}
+
+pub async fn send_rpc_request<T: Serialize>(
+    stream: &mut TcpStream, method: Method, args: &T,
+) -> LbResult<()> {
+    let method_id = method as u16;
+    let body = bincode::serialize(args).map_err(core_err_unexpected)?;
+    let len = body.len();
```

Comment:
I think sending a len prefix is fine and pretty common

you should just know about another approach, one that fits this 1 connection per request a bit better:

after finishing your writes you can execute `stream.Shutdown(write)` and then on the reading side you can do `read_all`. This means you don't have to negotiate the types.

This lets you do a single, probably more optimal syscall, signalling to the OS, I want to put all available bytes into memory, I assume this is the fastest way to express the idea if the end goal is in-fact to get all the bytes into memory.